### PR TITLE
use RecordBuilders in the Referrers plugin

### PR DESCRIPTION
**Pull request by diosmosis**
### Description:

**This PR is based off of #20720 which should be reviewed and merged first.**

Changes:
* Add RecordBuilder for Referrers.
* Add 'transform' property to Record that's only used for numeric records that are counts of blob records. This is to keep the logic in Referrers that does `$counts['recursive'] - $counts['level0']`. It doesn't seem like an ideal solution, and will be rarely used, but I couldn't come up with another.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/Referrers/RecordBuilders/Referrers.php [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    {
+        if (empty($row['referer_type'])) {
+            $row['referer_type'] = Common::REFERRER_TYPE_DIRECT_ENTRY;
+        }
+    }
+
+    /**
+     * @param DataTable[] $reports
+     */
+    protected function aggregateVisitRow($row, array $reports, array &$distinctUrls)
```
Comment (sgiehl) [inline plugins/Referrers/RecordBuilders/Referrers.php]:
```suggestion
    protected function aggregateVisitRow($row, array $reports, array &$distinctUrls): void
```

File: plugins/Referrers/RecordBuilders/Referrers.php [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    private function aggregateFromVisits(LogAggregator $logAggregator, array $reports, array &$distinctUrls, array $fields): void
+    {
+        $query = $logAggregator->queryVisitsByDimension($fields);
+        while ($row = $query->fetch()) {
+            $this->makeReferrerTypeNonEmpty($row);
+            $this->aggregateVisitRow($row, $reports, $distinctUrls);
+        }
+    }
+
+    protected function makeReferrerTypeNonEmpty(&$row)
```
Comment (sgiehl) [inline plugins/Referrers/RecordBuilders/Referrers.php]:
```suggestion
    protected function makeReferrerTypeNonEmpty(&$row): void
```

File: plugins/Referrers/RecordBuilders/Referrers.php [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_KEYWORD_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::KEYWORDS_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_CAMPAIGN_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::CAMPAIGNS_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_WEBSITE_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::WEBSITES_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_URLS_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::WEBSITES_RECORD_NAME, true)
+                ->setMultiplePeriodTransform(function (float $value, array $counts) {
+                    return $counts['recursive'] - $counts['level0'];
```
Comment (sgiehl) [inline plugins/Referrers/RecordBuilders/Referrers.php]: Might be good to add a comment here, so is easier to understand what exactly this is good for.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left some minor comments. Otherwise it seems some related test are failing...

**Comment by diosmosis** [comment]
@sgiehl updated

**Comment by sgiehl** [comment]
@diosmosis The ReportExporting UI tests are still failing. Seems some value changed from int to string or so. See https://github.com/matomo-org/matomo/actions/runs/5512669036/jobs/10049800824?pr=20740#step:3:1022

**Comment by diosmosis** [comment]
@sgiehl should be fixed

**Review by sgiehl — APPROVED** [review]
Looks good now 🎉

**Pull request merged**